File: algorithms-week9/bipartite/cpp/bipartite.cpp
//
// Created by Sam on 9/13/2018.
//

#include <iostream>
#include <vector>
#include <queue>

using std::vector;
using std::queue;

static int isBipartite = 1;

static void bfs(vector<vector<int>>& adj,
    vector<bool>& visited,
    vector<bool>& color,
    int s) {
  visited[s] = true;
  color[s] = true;
  queue<int> q;
  q.push(s);

  while (!q.empty()) {
    int v = q.front();
    q.pop();

    for (int& w : adj[v]) {
      if (!visited[w]) {
        visited[w] = true;
        color[w] = !color[v];
        q.push(w);
      } else if (color[v] == color[w]) {
        isBipartite = 0;
        return;
      }
    }
  }
}

int bipartite(vector<vector<int> > &adj) {
  size_t n = adj.size();
  vector<bool> visited(n);
  vector<bool> color(n);

  for (int v = 0; (v < n) && (isBipartite == 1); v++) {
    if (!visited[v]) {
      bfs(adj, visited, color, v);
    }
  }
  return isBipartite;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  std::cout << bipartite(adj);
}

File: algorithms-week9/bipartite/python/bipartite.py
#Uses python3

import sys
import queue

isBipartite = 1

def bipartite(adj):
    global isBipartite
    n = len(adj)
    visited = [False] * n
    color = [False] * n

    for v in range(n):
        if isBipartite == 0:
            break
        if visited[v] == False:
            bfs(adj, visited, color, v)

    return isBipartite

def bfs(adj, visited, color, s):
    global isBipartite
    q = queue.Queue()
    visited[s] = True
    color[s] = True

    q.put(s)

    while not q.empty():
        v = q.get()

        for w in adj[v]:
            if visited[w] == False:
                visited[w] = True
                color[w] = not color[v]
                q.put(w)
            elif color[v] == color[w]:
                isBipartite = 0
                return

    return

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(bipartite(adj))
